### PR TITLE
EVG-6628 Start and stop instances via the host spawn page

### PR DESCRIPTION
**Pull request by tzembo**
This adds pause and play buttons alongside the existing trash can that allows users to stop and start their instances using the Evergreen UI. The pause button only appears when the host is running, and the play button only appears when the host is stopped. 

Screen Shot 2019-10-01 at 11 13 24 AM

Screen Shot 2019-10-01 at 11 15 00 AM

### CODE REVIEW

**Comment by tzembo** [comment]
New yellow labels for stopping and stopped statuses:
Screen Shot 2019-10-01 at 1 18 44 PM

File: service/spawn.go [hunk 1]
```diff
@@ -212,7 +215,45 @@ func (uis *UIServer) modifySpawnHost(w http.ResponseWriter, r *http.Request) {
...
+			gimlet.WriteJSONError(w, fmt.Sprintf("Host %v is already stopping or stopped", h.Id))
+			return
+		}
+		if h.Status != evergreen.HostRunning {
+			gimlet.WriteJSONError(w, fmt.Sprintf("Host %v is not running", h.Id))
+			return
+		}
+
+		// Stop the host
+		queue := env.RemoteQueue()
```
Comment (tychoish) [inline service/spawn.go]: I'd use the uis.queue rather than going through the environment.

File: public/static/js/services/rest.js [hunk 2]
```diff
@@ -272,6 +272,24 @@ mciServices.rest.factory('mciSpawnRestService', ['mciBaseRestService', function(
...
+
+    service.startHost = function(action, hostId, data, callbacks) {
+        var config = {
+            data: data
+        };
+        config.data['action'] = action;
+        config.data['host_id'] = hostId;
+        baseSvc.postResource(resource, [], config, callbacks);
+    };
+
```
Comment (john-m-liu) [inline public/static/js/services/rest.js]: These can be the same function

File: public/static/js/spawned_hosts.js [hunk 3]
```diff
@@ -212,6 +212,34 @@ mciModule.controller('SpawnedHostsCtrl', ['$scope','$window', '$timeout', 'mciSp
...
+          success: function(resp) {
+            window.location.href = "/spawn";
+          },
+          error: function(jqXHR, status, errorThrown) {
+            notificationService.pushNotification('Error starting host: ' + resp.data.error,'errorHeader');
+          }
+        }
+      );
+    };
+
```
Comment (john-m-liu) [inline public/static/js/spawned_hosts.js]: These can probably the same function with the action as a param

File: public/static/js/spawned_hosts.js [hunk 4]
```diff
@@ -363,6 +393,29 @@ mciModule.controller('SpawnedHostsCtrl', ['$scope','$window', '$timeout', 'mciSp
           $scope.modalOpen = false;
           $scope.$apply();
         });
+      } else if ($scope.modalOption === 'stopHost') {
+        $scope.modalTitle = 'Stop Host';
+        modal.on('shown.bs.modal', function() {
+          $scope.modalOpen = true;
+          $scope.$apply();
```
Comment (john-m-liu) [inline public/static/js/spawned_hosts.js]: Do you specifically need to rerender the UI based on the data here? I wouldn't think so. Can you verify by removing this line and the other lines to $apply, and seeing if the escape key still dismisses the modal?

File: public/static/js/spawned_hosts.js [hunk 5]
```diff
@@ -363,6 +393,29 @@ mciModule.controller('SpawnedHostsCtrl', ['$scope','$window', '$timeout', 'mciSp
...
+        });
+
+        modal.on('hide.bs.modal', function() {
+          $scope.modalOpen = false;
+          $scope.$apply();
+        });
+
```
Comment (john-m-liu) [inline public/static/js/spawned_hosts.js]: I think all these options and the above can be combined into a single function

File: public/static/js/spawned_hosts.js [hunk 6]
```diff
@@ -384,6 +437,12 @@ mciModule.controller('SpawnedHostsCtrl', ['$scope','$window', '$timeout', 'mciSp
         if ($scope.modalOption === 'terminateHost') {
           $scope.terminateHost();
           $('#spawn-modal').modal('hide');
+        } else if ($scope.modalOption === 'stopHost') {
+          $scope.stopHost();
+          $('#spawn-modal').modal('hide');
+        } else if ($scope.modalOption === 'startHost') {
+          $scope.startHost();
+          $('#spawn-modal').modal('hide');
```
Comment (john-m-liu) [inline public/static/js/spawned_hosts.js]: These and the above can probably combined into a single switch-case on the action

File: public/static/js/spawned_hosts.js (hunk 4 above)
Comment (tzembo) [reply public/static/js/spawned_hosts.js]: Doesn't seem necessary! Removed them.

File: public/static/js/spawned_hosts.js [hunk 7]
```diff
@@ -332,62 +334,60 @@ mciModule.controller('SpawnedHostsCtrl', ['$scope','$window', '$timeout', 'mciSp
...
+      modal.on('shown.bs.modal', function() {
+        $scope.modalOpen = true;
+      });
+
+      modal.on('hide.bs.modal', function() {
+        $scope.modalOpen = false;
+      });
+    }
+
+    onEnterPressed = function(action) {
```
Comment (john-m-liu) [inline public/static/js/spawned_hosts.js]: This variable is named deceptively - when a function starts with "on," that implies that it's a function called when a certain event happens. In this case, the function actually attaches an event handler, so I would name this something like "attachEnterHandler"

Comment (tzembo) [reply public/static/js/spawned_hosts.js]: Good point! Fixed.

**Pull request merged**